blockchain-albatross/src/blockchain.rs

```diff
@@ -592,6 +599,39 @@ impl Blockchain {
                 return Err(PushError::InvalidBlock(BlockError::InvalidJustification));
             }
 
+            // Check if there are two blocks in the same slot and with the same height. Since we already
+            // verified the validator for the current slot, this is enough to check for fork proofs.
+            // Count the microblocks after the last macroblock
+            let micro_blocks: Vec<Block> = self.chain_store.get_blocks_at(block.block_number(), false, Some(&read_txn));
+
+            // Get the microheader from the block
+            let micro_header1 = &micro_block.header;
+
+            // Get the justification for the block. We asume that the
+            // validator's signature is valid.
+            let justification1 = &micro_block.justification.signature;
+
+            // Get the view number from the block
+            let view_number = block.view_number();
+
+            for micro_block in micro_blocks.iter() {
+                // If there's another microblock set to this view number, we
+                // notify the fork event.
+                if view_number == micro_block.view_number() {
+                    let test = micro_block.unwrap_micro();
+                    let micro_header2 = micro_block.clone().unwrap_micro().header;
+                    let justification2 = micro_block.clone().unwrap_micro().justification.signature;
```

Comment:
Ah, I think my previous suggestion was probably not working as `micro_block` is a reference.
To get rid of cloning the whole block twice here, you can do:
```suggestion
                    let micro_header2 = micro_block.unwrap_micro_ref().header.clone();
                    let justification2 = micro_block.unwrap_micro_ref().justification.signature.clone();
```
Or alternatively:
```suggestion
                    let micro_block = micro_block.clone().unwrap_micro();
                    let micro_header2 = micro_block.header;
                    let justification2 = micro_block.justification.signature;
```

In the first case, you're only cloning the parts you need, in the second you clone the whole block once.

And to have another alternative: The whole cloning is required, since we do `for micro_block in micro_blocks.iter() {` and `.iter()` gives us an iterator over references `&Block`.
However, since we don't need the `micro_blocks` for anything else afterwards, we could do:
`for micro_block in micro_blocks {` or even `for micro_block in micro_blocks.drain(..).map(|block| block.unwrap_micro()) {`.
The first will consume `micro_blocks` and iterate over `Block`. The second requires `micro_blocks` to be declared as `mut` as it empties the vector (plus I also did the conversion to `MicroBlock` in the `map()` here).

Reply:
I would actually choose one of the latter two possibilities as they avoid the cloning altogether. :)